File: nodejs/number-of-recent-calls933.js

var RecentCounter = function() {
    arr = [];
    arr.push(0);
};

/** 
 * @param {number} t
 * @return {number}
 */
RecentCounter.prototype.ping = function(t) {
    // console.log(arr);
    arr.push(t);
    if (t <= 3000) return arr.length-1;
    else {
        // console.log(t);
        let hi = arr.length;
        let lo = 0;
        while (lo!==hi) {
            
            let mid = Math.floor((lo+hi+1)/2);
            // console.log(mid + ' ' + lo + ' ' + hi);
            
            // if (arr[mid] <= t-3000) lo = mid;
            // else hi = mid - 1;
            if (arr[mid]>t-3000) hi = mid - 1;
            else lo = mid;
        }
        // console.log(arr);
        // console.log(`for ${t} lo is ${lo} ${t-3000 === arr[lo]}   ${arr.length} ${arr.length-lo+1}`);
        return (t-3000 === arr[lo]) ? arr.length-lo: arr.length - lo - 1;
    }

};




/** 
 * Your RecentCounter object will be instantiated and called as such:
 * var obj = Object.create(RecentCounter).createNew()
 * var param_1 = obj.ping(t)
 */


 var obj = new RecentCounter();
//  obj.ping(1);
//  obj.ping(2);

 console.log('res is: ' + obj.ping(1));
 console.log('res is: ' + obj.ping(2));
 console.log('res is: ' + obj.ping(100));
 console.log('res is: ' + obj.ping(3000));
 console.log('res is: ' + obj.ping(3001));
 console.log('res is: ' + obj.ping(3002));
 console.log('res is: ' + obj.ping(3003));
 console.log('res is: ' + obj.ping(4003));
 console.log('res is: ' + obj.ping(9002));
 

File: golang/numberofrecentcalls933/number-of-recent-calls933.go
package main

import "fmt"

func main() {
	fmt.Println("this")
	obj := Constructor()
	fmt.Println(obj.Ping(1))
	fmt.Println(obj.Ping(2))
	fmt.Println(obj.Ping(100))
	fmt.Println(obj.Ping(3000))
	fmt.Println(obj.Ping(3001))
	fmt.Println(obj.Ping(3002))
	fmt.Println(obj.Ping(3003))
	fmt.Println(obj.Ping(4003))
	fmt.Println(obj.Ping(9002))

}

//RecentCounter test asd
type RecentCounter struct {
	s1 []int
}

//Constructor fafa
func Constructor() RecentCounter {
	s1 := make([]int, 0)
	s1 = append(s1, 0)
	return RecentCounter{s1: s1}
}

//Ping fafa
func (e *RecentCounter) Ping(t int) int {
	e.s1 = append(e.s1, t)
	if t <= 3000 {
		return len(e.s1) - 1
	} else {
		lo := 0
		hi := len(e.s1) - 1
		for lo != hi {
			mid := (lo + hi + 1) / 2
			if e.s1[mid] > t-3000 {
				hi = mid - 1
			} else {
				lo = mid
			}
		}
		if e.s1[lo] == t-3000 {
			return len(e.s1) - lo

		} else {
			return len(e.s1) - lo - 1
		}
	}

}

/**
 * Your RecentCounter object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Ping(t);
 */
